Dialog/Panel: add onDismissed dialog event handler

scriby opened the pull request:
#### Pull request checklist

- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [ ] Documentation update

#### Description of changes

When a dialog is closed, it calls the onDismiss handler. When the dialog closing animation finishes, it calls the onDismiss handler again. It's difficult to distinguish whether the dialog has just begun closing or is done closing.

In this PR, I added an onDismissed callback which is called when the dialog is done being dismissed.

#### Focus areas to test

Dialog opening / closing. I didn't see any automated tests in this repo. Did I miss that or are there no tests?

msftclas commented:
Hi __@scriby__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, MSBOT;

dzearing commented:
I like it. However it would be nice to fix a few of the other cases where we have onDismiss; at least `Panel`. Would you mind adding it to Panel and doing a scan for any other place that has onDismiss to see if it applies?

scriby commented:
Sure, I'll work on those items.

scriby commented:
1. I implemented onDismissed in Panel as well. I didn't see any other components with a close animation.

2. I copied some other change files I saw, not exactly sure if I did it right. I tried following the instructions but rush just said something about not needing a change file.

3. Added some tests. I spent over an hour just trying to figure out a way to test it, so hopefully they're acceptable. Curious if there's a faster way to run tests as having to wait a few minutes each time to get feedback is a little painful.

scriby commented:
Oh... one other thing to point out is this is a breaking change. Not sure how/where to call that out.

dzearing commented on packages/office-ui-fabric-react/src/components/Dialog/Dialog.test.tsx:
```diff
@@ -0,0 +1,39 @@
+    const div = document.createElement('div');
+    ReactDOM.render(<Dialog isOpen={true} onDismissed={handleDismissed} />, div);
+      ReactDOM.render(<Dialog isOpen={false} onDismissed={handleDismissed} />, div);
+      const event = document.createEvent('CustomEvent'); // AnimationEvent is not supported by PhantomJS
```
nice work figuring out the tests!

dzearing commented on packages/office-ui-fabric-react/src/components/Panel/Panel.Props.ts:
```diff
@@ -33,11 +33,16 @@ export interface IPanelProps extends React.Props<Panel> {
 
   /**
-  * Event handler for when the panel is closed.
+  * A callback function for when the panel is closed.
```
Edit comment to clarify "when the panel will be closed, before the animation starts."

dzearing commented on packages/office-ui-fabric-react/src/components/Panel/Panel.Props.ts:
```diff
@@ -33,11 +33,16 @@ export interface IPanelProps extends React.Props<Panel> {
   onDismiss?: () => void;
 
   /**
+   * A callback function which is called after the Panel is dismissed.
```
...and the animation is complete.

dzearing commented on common/changes/scriby-dialog-on-dismissed_2017-01-31-19-40.json:
```diff
@@ -0,0 +1,10 @@
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Dialog: Call onDismissed callback function when dialog close animation finishes.",
+      "type": "patch"
```
Edit this (and the other change file) and change type to "minor". Minor fixes are non breaking, but introduce some new feature that is not available in previous patches.

Patches are for no prop change bug fixes. 

If the new param were not optional, or you were removing or significantly changing some behavior, that would be major. But I think you aren't doing that here.

scriby commented:
I was thinking about one more thing, I'll run it by you to see what you think before merging.

If the browser doesn't support animations or has animations disabled, the onDismissed event will never fire. It seems better if onDismissed fires immediately after onDismiss in that case.

We could try to feature detect whether animations are supported and vary the behavior.

msftclas commented:
Hi __@scriby__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

dzearing commented:
Hmm... seems the branch status on this change is hanging...trying to merge :)
